test/trusted_resources_test.go

```diff
@@ -0,0 +1,375 @@
+	}
+
+	secret := &v1.Secret{Data: map[string][]byte{"cosign.pub": fileBytes}, ObjectMeta: metav1.ObjectMeta{Name: "verification-secrets", Namespace: system.Namespace()}}
+	c.KubeClient.CoreV1().Secrets(system.Namespace()).Create(ctx, secret, metav1.CreateOptions{})
+	// Check if secret created
+	_, err = c.KubeClient.CoreV1().Secrets(system.Namespace()).Get(ctx, secret.Name, metav1.GetOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	configMapData = map[string]string{
+		config.PublicKeys: fmt.Sprintf("k8s://%s/verification-secrets", system.Namespace()),
+	}
+	if err := updateConfigMap(ctx, c.KubeClient, system.Namespace(), config.GetTrustedResourcesConfigName(), configMapData); err != nil {
+		t.Fatal(err)
+	}
+
+	// create pipelines
+	fqImageName := getTestImage(busyboxImage)
+	task := parse.MustParseTask(t, fmt.Sprintf(`
+metadata:
+  name: %s
+  namespace: %s
+spec:
+  steps:
+  - image: %s
+    command: ['/bin/sh']
+    args: ['-c', 'echo hello']
+`, helpers.ObjectNameForTest(t), namespace, fqImageName))
+
+	signedTask, err := GetSignedTask(task, signer, "signedtask")
+	if err != nil {
+		t.Errorf("error getting signed task: %v", err)
+	}
+	if _, err := c.TaskClient.Create(ctx, signedTask, metav1.CreateOptions{}); err != nil {
+		t.Fatalf("Failed to create Task: %s", err)
+	}
+
+	pipeline := parse.MustParsePipeline(t, fmt.Sprintf(`
+metadata:
+  name: %s
+  namespace: %s
+spec:
+  tasks:
+  - name: task
+    taskRef:
+      name: %s
+      kind: Task
+`, helpers.ObjectNameForTest(t), namespace, signedTask.Name))
+
+	signedPipeline, err := GetSignedPipeline(pipeline, signer, "signedpipeline")
+	if err != nil {
+		t.Errorf("error getting signed pipeline: %v", err)
+	}
+
+	tamperedPipeline := signedPipeline.DeepCopy()
+	tamperedPipeline.Name = "tampered"
+	tamperedPipeline.Annotations["foo"] = "bar"
+
+	tcs := []struct {
+		name     string
+		pipeline v1beta1.PipelineObject
+		wantErr  bool
+	}{{
+		name:     "Signed Pipeline Passes Verification",
+		pipeline: signedPipeline,
+		wantErr:  false,
+	}, {
+		name:     "Tampered Pipeline Fails Verification with tampered content",
+		pipeline: tamperedPipeline,
+		wantErr:  true,
+	}, {
+		name:     "Unsigned Pipeline Fails Verification without signature",
+		pipeline: pipeline,
+		wantErr:  true,
+	},
+	}
+
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+
+			if _, err := c.PipelineClient.Create(ctx, tc.pipeline.(*v1beta1.Pipeline), metav1.CreateOptions{}); err != nil {
+				t.Fatalf("Failed to create Pipeline: %s", err)
+			}
+
+			pr := parse.MustParsePipelineRun(t, fmt.Sprintf(`
+metadata:
+  name: %s
+  namespace: %s
+spec:
+  pipelineRef:
+    name: %s
+`, helpers.ObjectNameForTest(t), namespace, tc.pipeline.PipelineMetadata().Name))
+
+			t.Logf("Creating PipelineRun %s", pr.Name)
+			if _, err := c.PipelineRunClient.Create(ctx, pr, metav1.CreateOptions{}); err != nil {
+				t.Fatalf("Failed to create PipelineRun `%s`: %s", pr.Name, err)
+			}
+
+			if tc.wantErr {
```

Comment:
I recommend avoiding using the same test code to test error and success cases. it's a Tekton wide standard at  https://github.com/tektoncd/community/blob/main/standards.md#tests - would be nice if we included some of the reasoning for this in the docs itself but in the meantime you can see it in the discussion here https://github.com/tektoncd/community/pull/133#discussion_r451229109 - TL;DR: it makes code confusing and brittle and is much clearer and easier to maintain as separate test cases (even within this test to reason about what it's doing you have to understand that there are multiple `if tc.wantErr` blocks that completely change the test funcitonality)

Reply:
Sure! I started with having error and success cases separately but merge them to reduce the redundant code. I will divide them again

Reply:
Will keep this standard in mind!

Reply:
Fixed this in latest commit

Reply:
Thanks very much! (If there is code it makes sense for them to share, you could always consider creating a library for that code - but also a little repetition is often not as bad as it seems - 'repetition is better than the wrong abstraction' :D (https://sandimetz.com/blog/2016/1/20/the-wrong-abstraction)